Runtime Storage

danforbes opened the pull request:
Closes #5

danforbes committed:
[WIP] Runtime storage

Closes #5

danforbes committed:
Add section on calculating storage keys and using RPC to query storage

danforbes committed:
Use 120-character line length

danforbes committed:
Document methods for Storage Items

danforbes committed:
Small changes to runtime storage document

danforbes commented:
@thiolliere can you please review this?

gui1117 commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,128 @@ lang: en
+the size of the object will begin to incur costs that may outweigh the optimization in storage reads. Storage values
+can be used to store lists of items, but runtime developers should take care with respect to the size of these lists.
+Large lists incur storage costs just like large `structs`. Furthermore, iterating over a large list in your runtime may
+result in exceeding the block production time.
```
Yes but I guess if your list is too long to be iterated upon, then just decoding it would be already very very long. I think the emphasize on the cost to read/write it is more important. but ok anyway

danforbes replied:
I believe that Shawn has indicated we should specifically caution people against iteration within the runtime.

gui1117 replied:
(ok but just to clarify if you store in your list something different from basic u32, u64... then decoding/encoding is done by iterating other elements.)

danforbes replied:
Can you please expand on that concept just a bit for me? Are you referring to the fact that anything other than a Rust primitive requires a pointer to be dereferenced?

gui1117 replied:
in parity-scale-codec to decode `Vec<u32>` and all other type in `	U8, I8, U16, I16, U32, I32, U64, I64, U128, I128,` we decode/encode by doing mem::transmute, otherwise we do:
```rust
fn encode(..) {
...
				for item in self {
					item.encode_to(dest);
				}
}
```
and 
```rust
fn decode(...) {
...
					for _ in 0..len {
						r.push(T::decode(input)?);
					}
}
```

danforbes replied:
I see, so complex data types require an extra layer of indirection that materializes as an encoding/decoding step. Correct?

gui1117 replied:
I wouldn't phrases this way. Just encoding/decoding of vec is optimized for `U8, I8, U16, I16, U32, I32, U64, I64, U128, I128` I think also Option is optimized for them and some other generic data type.
But this is just optimization in my point of view.

danforbes replied:
Do you think that it is important for us to document this optimization with respect to the runtime storage interface?

gui1117 replied:
not sure, should belong to some codec doc I think

danforbes replied:
Opened a separate issue :+1: #43

gui1117 commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,128 @@ lang: en
+* Iterable Storage Maps - Implementations of
+  [Storage Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageMap.html) and
+  [Storage Double Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageDoubleMap.html)
+  whose keys and values can be iterated over
```
I would have actually thought that we would present 3 kind of storage: value, map and double_map and map and double_map are iterable when they use hasher with key information.

like in my mind they declare a map storage and this map storage implements StorageMap and optionnally IterableStorageMap traits.

I think it is less confusing presented this way, but I have no strong opinion

danforbes replied:
I would be interested to hear the opinions of other people on this topic. I am trying to approach this document from the POV of the capabilities that users will be consuming, which is why I have chosen to describe things the way that I do.

shawntabrizi replied:
To our users, there are 3 types of storage: value, map, double_map.

The traits on top of these storage are opaque to the user, so I agree with @thiolliere

gui1117 commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,128 @@ lang: en
+
+##### `take()`
+
+Remove the value from storage.
```
maybe the name of those method are self explanatory and should just directly link to the inner doc? (not sure)

danforbes replied:
I tried to take a best-of-both-worlds approach by deemphasizing these sections within the document. I think it's good context to provide, but agree that the Rust docs should be the canonical source with respect to this kind of stuff.

gui1117 commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,128 @@ lang: en
+
+### Storage Maps
+
+Storage Maps are implemented as hash maps, which is a pattern that should be familiar to most developers. In order to
```
> Storage Maps are implemented as hash maps

maybe it is only me but for me hash map is a data structure which handle collision. at least if presented this way it should be clear that collision are not handled.

But actually instead of using "hash map" term I would rather say that it is a map where key are hashed.

danforbes replied:
Thank you for this clarification. Please let me know if you think my solution was sufficient.

gui1117 replied:
hmm I see, I'm not sure actually what I proposed was better, like

> Storage Maps are implemented as maps with hashed keys, which is a pattern that should be familiar to most developers.

wouldn't have make sense to me I think if I didn't know it already, not sure how to do better. (still for me the clearest is that a storage map insert at the path `hash(encode(key))` the value encoded.)

danforbes replied:
Tried to clarify this in the docs :+1:

gui1117 commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,128 @@ lang: en
+Storage Maps are implemented as hash maps, which is a pattern that should be familiar to most developers. In order to
+give blockchain engineers increased control over the way in which these data structures are stored, Substrate allows
+developers to select the hashing algorithm that is used to generate map keys. Map data structures are ideal for managing
+sets of items whose elements will be accessed randomly, as opposed to iterating over them sequentially in their entirety.
```
I think it would be better to show how it is implemented in the underlying trie to make it more understandable about advantage/disadvantage.

Basically the value is put in the trie at the path==some_storage_prefix++hashed_key.
So each value is stored in an individual node, so reading/writing a value only decode/encode one raw_value which is stored in one trie-node.

also the performance of the map will depends on how big the map is because read/writing one value will recompute the path==some_prefix++hashed_key. and the hashed_key part can be more or less costy to compute. https://www.shawntabrizi.com/substrate/transparent-keys-in-substrate/

danforbes replied:
There is [a separate document that goes over how storage is implemented](https://github.com/substrate-developer-hub/knowledgebase/blob/runtime-storage/current/advanced/storage.md). This document is meant to explain Substrate's storage interfaces and help blockchain runtime developers understand when and how to use them.

gui1117 replied:
this document will not explain how frame_support::storage abstraction are implemented in storage do they ? so I would explain their implementation here actually.

danforbes replied:
I believe the purpose of the advance document is to cover implementation and the purpose of this document (the runtime document) is to cover the interface. @joepetrowski can you provide your input?

gui1117 commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,128 @@ lang: en
+
+[Storage Maps expose an API](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#required-methods)
+that is similar to that of Storage Values. The selected methods referenced below all use a single key; use two keys for
+[Storage Double Maps](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#required-methods):
```
Here also I think we should say that using this two keys result in this in the trie: key1, key2, value is stored at
value is stored at some_prefix++hashe1(key1)++hasher2(key2).
and so you can iterate on one key1.

danforbes replied:
This is information I also think would belong in the advanced document.

gui1117 commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,128 @@ lang: en
+The Substrate storage API provides iterable map implementations. Because maps are often used to track unbounded sets of
+data (account balances, for example) it is especially likely to exceed block production time by iterating over maps in
+their entirety within the runtime. Furthermore, because maps are comprised of more layers of indirection than native
+lists, they are significantly more costly than lists to iterate over with respect to time. Depending on
```
I think we can just explain what is the "more layers of indirection than native lists", basically iterating over a map is just iterating over all keys after the prefix `storage_prefix` so this trie iteration is more costy than native list because it is a trie.

danforbes replied:
I felt like that was a step too far for this document and that topics like that belong in [the advanced storage documentation](https://github.com/substrate-developer-hub/knowledgebase/blob/runtime-storage/current/advanced/storage.md).

gui1117 replied:
The above document just explain how the trie works not how frame_support storage abstractions are implemented using the trie, no ?
As a user in order to understand more precisely the cost of those abstraction, explaining their implementation on the trie seems useful (maybe it belong somewhere else in the doc but still I feel like saying that a storage map just store all its value after a prefix and iterating upon means iterating the trie after a prefix is quite straightforward and quickly understandable)

> Because maps are often used to track unbounded sets of
data it is especially likely to exceed block production time by iterating over maps in
their entirety within the runtime

This is just a trade off, some storage map in the runtime are quite bounded actually. But we use storage map because they are more often accessed by individual value than as a whole. Thus using a storage map is more efficient because accessing one value is cheaper, but it is sometime iterated upon still because it is not too big.

danforbes replied:
Tried to clarify this in the docs :+1:

gui1117 commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,128 @@ lang: en
+their entirety within the runtime. Furthermore, because maps are comprised of more layers of indirection than native
+lists, they are significantly more costly than lists to iterate over with respect to time. Depending on
+[the hashing algorithm](#Transparent-Hashing-Algorithms) that you select to generate a map's keys, you may be able to
+iterate across its keys as well as its values.
```
Maybe we should presice here that IterableStorageMap is only implemented for Map if the hasher is reversible. (i.e. implements ReversibleStorageHasher, which is not shown in the doc but should be IMO)

danforbes replied:
Is `ReversibleStorageHasher` documented anywhere in the Rust docs? I only see it mentioned in [the docs for `IterableStorage(Double)Map`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageMap.html#implementors) but it's not indexed so difficult to link to.

gui1117 replied:
it is in the doc since https://github.com/paritytech/substrate/pull/5504 should be soon in the generated doc I expect. I agree it was a probably mistake not to expose it.

gui1117 replied:
but thinking again it seems that non reversible hasher are deprecated so maybe not worth mentionning:
```
/// Deprecated hashers, which do not support iteration over keys include:
/// * `twox_128` - TwoX with 128bit.
/// * `twox_256` - TwoX with with 256bit.
/// * `blake2_128` - Blake2 with 128bit.
/// * `blake2_256` - Blake2 with 256bit.
```

gui1117 commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,128 @@ lang: en
+#### Methods
+
+[Iterable Storage Maps expose the following methods](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageMap.html#required-methods)
+in addition to the other map methods:
```
their method are slightly different for map and doublemap

danforbes replied:
Tried to clarify this in the docs :+1:

gui1117 commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,128 @@ lang: en
+##### `translate(fn)`
+
+Use the provided function to translate all elements of the map, in no particular order. To remove an element from the
+map, return `None` from the translation function.
```
they also differ between map and double_map (as a note I tried to unify interface here https://github.com/paritytech/substrate/pull/5335)

danforbes replied:
Tried to clarify this in the docs :+1:

gui1117 reviewed (commented):
I feel like explaining what each methods do in the trait is a bit redundant with the trait documentation.
I'm not sure of the scope of this doc but I feel like part of it is redundant with decl_storage and traits doc.

Also in my opinion storage doc should be more about explaining the trie and then just saying:
* value is an abstraction which stores one value at a certain path in the trie
* map stores value at some_path++hasher(key)
* double map stores avalue at some_path++ hasher1(key)++hasher2(key)

I think then it is more understandable what are the cost of methods.

But I'm not very aware about how is organized developer-hub so this is just humble suggestions.

by the way I think Shawn deep dive was very good explanation of trie https://www.shawntabrizi.com/substrate/substrate-storage-deep-dive/

danforbes committed:
Address @thiolliere PR comments

* Deemphasize redudancy with Rust docs by collapsing sections about Storage Item APIs
* Distinguish APIs for StorageMap and StorageDoubleMap
* Clarify StorageMap implementation

danforbes commented:
Thank you @thiolliere! I have tried to address your comments while balancing the goal of this document with respect to the goal of [the advanced storage document](https://github.com/substrate-developer-hub/knowledgebase/blob/runtime-storage/current/advanced/storage.md).

danforbes committed:
Address @thiolliere PR comments

* Clarify Storage Map implementation
* Clarify iterable storage map principals

danforbes committed:
Formatting

danforbes committed:
Clarify the `get` extension to `decl_storage`

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,123 @@ lang: en
+consensus about the state of the underlying storage. Furthermore, well designed storage systems reduce the load on
+nodes in the network, which will lower the overhead for participants in your blockchain. Substrate exposes a set of
+layered, modular storage APIs that allow runtime developers to make the storage decisions that suit them best. However,
+the fundamental principal of blockchain runtime storage is to minimize its use. This document is intended to provide
```
I think this should be principle by definition 1c
https://www.merriam-webster.com/dictionary/principle

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,123 @@ lang: en
 ## Storage Items
 
-Your runtime module has access to Substrate storage APIs which allows you to easily store common storage items:
+[The `storage` module in the Substrate Support pallet](https://substrate.dev/rustdocs/master/frame_support/storage/index.html)
```
Support is not a pallet
```suggestion
The `storage` module in [FRAME Support](https://substrate.dev/rustdocs/master/frame_support/storage/index.html)
```

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,123 @@ lang: en
 
-Your runtime module has access to Substrate storage APIs which allows you to easily store common storage items:
+[The `storage` module in the Substrate Support pallet](https://substrate.dev/rustdocs/master/frame_support/storage/index.html)
+gives runtime developers access to Substrate's flexible storage APIs. Any value which can be encoded by the
```
```suggestion
gives runtime developers access to Substrate's flexible storage APIs. Any value that can be encoded by the
```

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,123 @@ lang: en
 
-You can use the `decl_storage!` macro to easily create new runtime storage items. Here is an example of what it looks like to declare each type of storage item:
+This type of storage item should be used for values that are viewed as a single unit by the runtime, whether that is a
+single primitive value, a single `struct` or a single collection of related items. Although wrapping related items in
```
```suggestion
single primitive value, a single `struct`, or a single collection of related items. Although wrapping related items in
```

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,123 @@ lang: en
+Storage Maps are implemented as maps with hashed keys, which is a pattern that should be familiar to most developers.
+Unlike traditional hash maps, though, [Storage Maps in Substrate are simple key-value stores](../advanced/storage) that
+do not take key collision into account. The hashing algorithms that Substrate supplies are designed so that runtime
+developers do not need to worry about key collisions, but the implementation of Substrate Storage Maps does becomes
```
```suggestion
developers do not need to worry about key collisions, but the implementation of Substrate Storage Maps does become
```

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,123 @@ lang: en
+data (account balances, for example) it is especially likely to exceed block production time by iterating over maps in
+their entirety within the runtime. Furthermore, because maps are comprised of more layers of indirection than native
+lists, they are significantly more costly than lists to iterate over with respect to time. This is not to say that it
+is "wrong" to iterate over maps in your runtime; in general Substrate focuses on "first principals" as opposed to hard
+and fast rules of right and wrong. Being efficient within the runtime of a blockchain is an important first principal
```
This one is definitely principle
```suggestion
is "wrong" to iterate over maps in your runtime; in general Substrate focuses on "first principles" as opposed to hard
and fast rules of right and wrong. Being efficient within the runtime of a blockchain is an important first principle
```

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -4,22 +4,123 @@ lang: en
+is "wrong" to iterate over maps in your runtime; in general Substrate focuses on "first principals" as opposed to hard
+and fast rules of right and wrong. Being efficient within the runtime of a blockchain is an important first principal
+of Substrate and this information is designed to help you understand _all_ of Substrate's storage capabilities and use
+them in a way that respects the important first principals around which they were designed. Depending on
```
```suggestion
them in a way that respects the important first principles around which they were designed. Depending on
```

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -70,6 +358,13 @@ TODO
 
-* Visit the reference docs for the [`decl_storage!` macro](https://substrate.dev/rustdocs/master/frame_support/macro.decl_storage.html) more details possible storage declarations.
+* Visit the reference docs for the
+  [`decl_storage!` macro](https://substrate.dev/rustdocs/master/frame_support/macro.decl_storage.html) more details
```
```suggestion
  [`decl_storage!` macro](https://substrate.dev/rustdocs/master/frame_support/macro.decl_storage.html) for more details about
```

joepetrowski reviewed (commented):
Looks good. I think a practical section on best practices would be useful.

shawntabrizi commented:
Is the goal of this PR to actually have all this content on a single page? Or are you just collecting information here?

Jumping between conceptual items around storage, to macro syntax about declaration, to RPC call stuff seems like a lot for one page, and what I would consider to be quite different topics when it comes to storage.

I might tune your focus into two sections with somewhat different goals:

* Conceptually what is Runtime Storage in Substrate?
   * Goal: Someone will understand from a theoretical perspective how substrate storage works, and the design decisions.

* Using storage in Substrate runtime code
   * Goal: How do work with the macro, the generated types, etc... what are all the nuiances of the syntax, things you can do, cant do etc...

Someone who does not really want to make the best decisions but just wants code to work should read the second doc. Someone who just wants to learn conceptually whats going on should read the first. A "real dev" would read both and be able to make the right decisions using the combined information.

If you break down Substrate Storage Conceptually, then a topic like the RPC queries become a lot more natural "We are just querying a database key, no other abstractions of storage leak through the RPC".

It also informs the user the advantages of a single value storage item versus a storage map, something that is probably the most important topic to take away for the average dev. They should be thinking at this level.

Then when it comes to the ergonomics of the language, some users just want to know what to type to make the thing compile, and that would go into all the syntax level details I see here.

As it is written, I think it is hard to ingest and get a clear message imo.

danforbes committed:
Clarifications and improvements from @joepetrowski

danforbes committed:
Remove iterable storage maps from top-level listing of storage items

Per suggestion of @thiolliere and @shawntabrizi

danforbes committed:
Rework Iterable Storage Map documentation

danforbes committed:
Structural changes per @shawntabrizi

danforbes committed:
Structural changes per @shawntabrizi (advanced storage)

danforbes committed:
Small changes for readability and elaboration

danforbes commented:
@thiolliere and @shawntabrizi - I have made some structural changes that I hope will address your comments. @joepetrowski - per our conversation yesterday, I think this is now ready for a final review.

danforbes committed:
Prettier

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -4,49 +4,227 @@ lang: en
+- [`mutate(fn)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#tymethod.mutate) -
+  Mutate the value with the provided function.
+- [`take()`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#tymethod.take) -
+  Remove the value from storage.
```
```suggestion
  Load the value and remove it from storage.
```

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -4,49 +4,227 @@ lang: en
+- `mutate` - Use the provided function to mutate the value associated with the given key. Docs:
+  [`StorageMap#mutate(key, fn)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#tymethod.mutate),
+  [`StorageDoubleMap#mutate(key1, key2, fn)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.mutate)
+- `take` - Remove the value associated with the given key from storage. Docs:
```
```suggestion
- `take` - Load the value associated with the given key and remove it from storage. Docs:
```

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -4,49 +4,227 @@ lang: en
+map's keys, you may be able to iterate across its keys and values. Because maps are often used to
+track unbounded sets of data (account balances, for example) it is especially likely to exceed block
+production time by iterating over maps in their entirety within the runtime. Furthermore, because
+maps are comprised of more layers of indirection than native lists, they are significantly more
```
What are layers of indirection?

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -4,49 +4,227 @@ lang: en
+- `translate` - Use the provided function to translate all elements of the map, in no particular
+  order. To remove an element from the map, return `None` from the translation function. Docs:
+  [IterableStorageMap#translate(fn)](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageMap.html#tymethod.translate),
+  [IterableStorageDoubleMap#translate(fn)](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageDoubleMap.html#tymethod.translate)
```
Should have \` \` around the linked items.

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -4,49 +4,227 @@ lang: en
+| [Identity](https://substrate.dev/rustdocs/master/frame_support/struct.Identity.html)                  |               |             |
+
+The Identity hasher encapsulates a hashing algorithm that has an output equal to its input (the
+identity function).This type of hasher should only be used when the starting key is already a
```
```suggestion
identity function). This type of hasher should only be used when the starting key is already a
```

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -4,49 +4,227 @@ lang: en
+		pub SomePrimitiveValue get(fn some_primitive_value): u32;
+		// complex types are prefaced by T::
+		pub SomeComplexValue: T::AccountId;
+		pub SomeMap get(fn some_map): map hasher(blake2_128_concat) str => u32;
```
We recommend against strings in the runtime, so I would change this type.

joepetrowski replied:
Could also add in a private storage item and explain its accessibility.

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -56,38 +234,61 @@ decl_storage! {
 TODO
 
-## Child Storage Tries
+## Best Practices
```
Let's add:

- Verify first, write last
- Store only consensus critical info on chain (use example of submitting preimage)
- Set limits to the length of vectors that users can affect

joepetrowski commented on current/advanced/storage.md:
```diff
@@ -64,11 +66,131 @@ can use to verify the specific content in that trie. Subsections of a trie do no
+To calculate the key for a simple [Storage Value](../runtime/storage#Storage-Value), take the
+[TwoX 128 hash](https://github.com/Cyan4973/xxHash) of the name of the module that contains the
+Storage Value and append to it the TwoX 128 hash of the name of the Storage Value itself. For
+example, the [Sudo](https://substrate.dev/rustdocs/master/pallet_sudo/index.html) module exposes a
```
```suggestion
example, the [Sudo](https://substrate.dev/rustdocs/master/pallet_sudo/index.html) pallet exposes a
```

joepetrowski commented on current/advanced/storage.md:
```diff
@@ -64,11 +66,131 @@ can use to verify the specific content in that trie. Subsections of a trie do no
-The Substrate's [Support module](https://substrate.dev/rustdocs/master/frame_support/index.html)
-provides utilities to generate unique, deterministic keys for your runtime module storage items.
-These storage items are placed in the state trie and are accessible by querying the trie by key.
+The Substrate [FRAME Support pallet](https://substrate.dev/rustdocs/master/frame_support/index.html)
```
```suggestion
Substrate's [FRAME Support](https://substrate.dev/rustdocs/master/frame_support/index.html)
```

danforbes committed:
Add section on best practices and address suggestions from @joepetrowski

danforbes commented:
@joepetrowski - any suggestions for enhancing [the section on genesis configuration](https://github.com/substrate-developer-hub/knowledgebase/blob/a54712a0b2502fd96e93f951362501168473ffaf/current/runtime/storage.md#genesis-config)?

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -56,38 +244,129 @@ decl_storage! {
+The state of a blockchain network's storage is immutable; data can be changed, but there will always
+be a record of these changes, and making them typically incurs costs. Because of this, it is
+important that data is only persisted to runtime storage when it is certain that all preconditions
+have been met. In general, code blocks that may result in adding data to storage should be
```
```suggestion
have been met. In general, code blocks that may result in mutating storage should be
```

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -56,38 +244,129 @@ decl_storage! {
+requires users to configure a bound on this number that will be included as a
+[precondition](#Verify-First-Write-Last) before anything is written to storage.
 
 ## Child Storage Tries
 
 TODO
 
+## Storage Cache
+
+TODO
+
```
This may be good in the Advanced doc but don't think it's needed here. TBH I'm not sure how much runtime developers even need to know about these two things.

joepetrowski commented on current/runtime/storage.md:
```diff
@@ -56,38 +244,129 @@ decl_storage! {
-## Verify First, Write Last
+### Genesis Config
 
-TODO
+You can define
+[an optional `GenesisConfig`](https://substrate.dev/rustdocs/master/frame_support/macro.decl_storage.html#genesisconfig)
+struct in order to initialize Storage Items in the genesis block of your blockchain.
 
-## Storage Cache
+// TODO
```
@thiolliere what info do you think should go here?

gui1117 replied:
GenesisConfig is generated by decl_storage macro, it does initialize storage items for the genesis block of the blockchain.

Its logic is defined by:
* `config` and `build` attribute when declaring an individual storage.
* add_extra_genesis informations. which allows to add fields in the GenesisConfig struct and a build function to build storages using those fields.

(I can be more precise actually we execute first for each storage the build closure if it exist, or if no build attribute but a config attribute then we add into storage the value found in config. Then we execute the add_extra_genesis::build function.)

danforbes committed:
Small clarifications

kianenigma commented on current/runtime/storage.md:
```diff
@@ -4,49 +4,237 @@ lang: en
+Depending on [the hashing algorithm](#Transparent-Hashing-Algorithms) that you select to generate a
+map's keys, you may be able to iterate across its keys and values. Because maps are often used to
+track unbounded sets of data (account balances, for example) it is especially likely to exceed block
+production time by iterating over maps in their entirety within the runtime. Furthermore, because
```
> Furthermore, because
accessing the elements of a map requires more pointer dereferencing than accessing the elements of a
native list, maps are significantly _more_ costly than lists to iterate over with respect to time.

Can you explain this a bit more? I am not sure where it comes from and until now my understanding of the state was that the most important factors encoding + db access. In-memory dereferencing seems negligible to me.

kianenigma replied:
(I agree that iterating maps is expensive, I am not sure if your reason is correct.)

danforbes replied:
So, this line used to read as follows:
> Furthermore, because maps are comprised of more layers of indirection than native lists...

I agree with you that the change didn't necessarily make things more clear. I guess what I meant above was that you don't need to dereference _memory_ pointers, per se, but you do have to dereference the pointer to the element in the map...I think I'm trying to describe the "db access" part that you mention above. I will think about alternate ways to phrase this but would definitely appreciate any suggestions :+1:

danforbes replied:
Okay, this now reads:
> ...accessing the elements of a map requires more database reads than accessing the elements of a native list...

I hope you agree the simplest approach was the best one in this case :sweat_smile:

kianenigma commented on current/runtime/storage.md:
```diff
@@ -4,49 +4,237 @@ lang: en
+wrong. Being efficient within the runtime of a blockchain is an important first principle of
+Substrate and this information is designed to help you understand _all_ of Substrate's storage
+capabilities and use them in a way that respects the important first principles around which they
+were designed.
```
would be nice to mention here when and which maps are iterable at the end of the day.

danforbes replied:
This is a great point and I'm starting to wonder if we should even document non-transparent hashers since they are now deprecated. Should we just behave as if all hashers are transparent and all maps are iterable?

danforbes replied:
Asked @shawntabrizi about this offline and he agrees that we should take the simpler approach: only document that which is not deprecated :+1:

danforbes replied:
Tried to clarify this in the docs :+1:

kianenigma commented on current/runtime/storage.md:
```diff
@@ -4,49 +4,237 @@ lang: en
+cryptographic hash algorithm this would create an imbalanced storage structure that would suffer in
+performance. Cryptographic hashing algorithms are more complex and resource-intensive than their
+non-cryptographic counterparts, which is why Substrate allows developers to select when they are
+used.
```
would be good to name the most common hashes here and say which is secure and which is not.

kianenigma commented on current/runtime/storage.md:
```diff
@@ -4,49 +4,237 @@ lang: en
+In the example above, all the storage items except `SomePrivateValue` are made public by way of the
+`pub` keyword. Blockchain storage is always publicly
+[visible from _outside_ of the runtime](#Accessing-Storage-Items); the visibility of Substrate
+storage items only impacts whether or not other runtime pallets will be able to access the storage
```
```suggestion
storage items only impacts whether or not other pallets _within the runtime_ will be able to access the storage
```

kianenigma reviewed (commented):
I only had a skim and the text reads ver well to me. As Shawn mentioned I think this is _a lot_ of info and should be organised well so the user can follow it easily. Great work 👍

thiolliere commented:
Sorry I didn't realized earlier but it seems non-transparent hasher are deprecated so:

> /// Supported hashers (ordered from least to best security):
///
/// * `identity` - Just the unrefined key material. Use only when it is known to be a secure hash
///   already. The most efficient and iterable over keys.
/// * `twox_64_concat` - TwoX with 64bit + key concatenated. Use only when an untrusted source
///   cannot select and insert key values. Very efficient and iterable over keys.
/// * `blake2_128_concat` - Blake2 with 128bit + key concatenated. Slower but safe to use in all
///   circumstances. Iterable over keys.
///
/// Deprecated hashers, which do not support iteration over keys include:
/// * `twox_128` - TwoX with 128bit.
/// * `twox_256` - TwoX with with 256bit.
/// * `blake2_128` - Blake2 with 128bit.
/// * `blake2_256` - Blake2 with 256bit.
///

Thus I would say write that storage map are iterable and double map are also iterable, and maybe add a note that this is not the case if you use deprecated hasher `opaque_twox_128`, `opaque_...` ...

danforbes committed:
Storage item is not a proper noun

danforbes committed:
Use crates.parity.io for Rustdoc links

danforbes committed:
Genesis configuration

danforbes committed:
Small enhancements per @kianenigma

danforbes committed:
Non-transparent hashers are deprecated

danforbes commented:
@shawntabrizi, @thiolliere, @kianenigma - I have no more changes to make at this time based on your comments. Let me know if I addressed your concerns properly or if there are additional concerns you have.

gui1117 commented on current/runtime/storage.md:
```diff
@@ -4,49 +4,240 @@ lang: en
+| [TwoX 64 Concat](https://crates.parity.io/frame_support/struct.Twox64Concat.html)          |               | X           |
+| [Identity](https://crates.parity.io/frame_support/struct.Identity.html)                    |               |             |
+| [Blake2 128](https://crates.parity.io/frame_support/struct.Blake2_128.html) **DEPRECATED** | X             |             |
+| [TwoX 128](https://crates.parity.io/frame_support/struct.Twox128.html) **DEPRECATED**      |               |             |
```
there is actually 4 depracated hasher but I don't mind if we don't mention them:



 >   twox_128 - TwoX with 128bit.
    twox_256 - TwoX with with 256bit.
    blake2_128 - Blake2 with 128bit.
    blake2_256 - Blake2 with 256bit.

gui1117 commented on current/runtime/storage.md:
```diff
@@ -4,49 +4,240 @@ lang: en
-		pub SomeDoubleMap: double_map u64, blake2_256(u64) => u64;
+		SomePrivateValue: u32;
+		pub SomePrimitiveValue get(fn some_primitive_value): u32;
+		// complex types are prefaced by T::
+		pub SomeComplexValue: T::AccountId;
```
I feel like the wording shows that there is macro magic in the syntax here. but actually this is a regular rust type.
(The macro parse the type themself to know if there are generic over `T`, if they are then the storage struct generated is generic as well.)

```suggestion
		// types can make use of generic `T: Trait`
		pub SomeComplexValue: T::AccountId;
```

gui1117 commented on current/runtime/storage.md:
```diff
@@ -56,13 +247,240 @@ decl_storage! {
+[the `sc_chain_spec::ChainSpec` trait](https://crates.parity.io/sc_chain_spec/trait.ChainSpec.html).
+For a complete and concrete example of using Substrate's genesis storage configuration capabilities,
+refer to the `decl_storage` macro in
+[the Society pallet](https://github.com/paritytech/substrate/blob/master/frame/society/src/lib.rs)
```
should it be better to point to example pallet ?

danforbes replied:
I don't believe that the example pallet does any genesis configuration.

gui1117 commented on current/runtime/storage.md:
```diff
@@ -56,13 +247,240 @@ decl_storage! {
+
+  // all storage writes go here; no throwing code below this line
+
+  // all event emissions go here
```
events can be interleaved with storage writes

gui1117 reviewed (commented):
looks good to me

gui1117 commented on current/runtime/storage.md:
```diff
@@ -56,13 +247,240 @@ decl_storage! {
+initializing storage, all of which have entry points in the `decl_storage` macro. These mechanisms
+all result in the creation of a `GenesisConfig` data type that implements
+[the `sp_runtime::BuildModuleGenesisStorage` trait](https://crates.parity.io/sp_runtime/trait.BuildModuleGenesisStorage.html)
+and will be added to the storage item's module (e.g.
```
I'm not sure I read this correctly, what does the runtime GenesisConfig being added to the storage item's module means ?

Maybe we could make it more clear that there is a GenesisConfig overarching type for the runtime which is made by construct_runtime by joining the specified pallet genesis config (specified with `Config`)

Commit:
Clarify generic storage types

Co-authored-by: thiolliere <gui.thiolliere@gmail.com>

danforbes committed:
Clarifications per review by @thiolliere

joepetrowski committed:
final nits

joepetrowski reviewed (approved):
:rocket:
